### ISSUE TITLE
🐛 [cloud_firestore] writing DocumentReference with type safe Type in Unhandled Exception.

### ISSUE THREAD

**Issue by devaronius** [FirebaseExtended/flutterfire#6369]
## Bug report

**Describe the bug**
When you get a DocumentReference from a Collection or document that uses a `withConverter` function or define the DocumentReference with a type for example  DocumentReference<Map<String, dynamic>> will result in a 

`[ERROR:flutter/lib/ui/ui_dart_state.cc(199)] Unhandled Exception: Invalid argument: Instance of '_WithConverterDocumentReference<Map<String, dynamic>>'`.

You can only write a DocumentReference when T has not been set.

### Steps to reproduce

1. define DocumentReference<Map<String,dynamic>> reference in a JSON map
2 a.  add to the collection as a document.
2 b. add or update the document with given  DocumentReference

3. see the error in the console.

### Expected behavior

The typesafe DocumentReference is saved into Firestore as a DocumentReference value inside your document.

### Sample project

class Dto {
  const Dto(
    this.reference,
  );

  final DocumentReference<Map<String, dynamic>>   reference;

  Map<String, dynamic> toJson() => {
      'reference': reference,
  }
}

final reference =  store.collection('test').withConverter<Map<String,dynamic>>(
        fromFirestore: (json, _) => json,
        toFirestore: (dto, _) => const {}).doc('value').doc();

final to =  Dto(reference);

store.collection('test').add(to);

---

## Additional context

none

---

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click To Expand</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.2.1, on macOS 11.4 20F71 darwin-x64, locale en-NL)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] Connected device (2 available)

```

</details>

---

### Flutter dependencies

Run `flutter pub deps -- --style=compact` and paste the output below:

<details><summary>Click To Expand</summary>

```
Dart SDK 2.13.1
Flutter SDK 2.2.1
pharmacy 0.14.1+2021060901

dependencies:
- android_play_install_referrer 0.1.1 [flutter]
- app_tracking_transparency 2.0.1 [flutter]
- camera 0.8.1+3 [camera_platform_interface flutter pedantic quiver]
- chopper 4.0.1 [http meta logging]
- connectivity_plus 1.0.2 [flutter meta connectivity_plus_platform_interface connectivity_plus_linux connectivity_plus_macos connectivity_plus_web connectivity_plus_windows]
- dartz 0.10.0-nullsafety.2
- datetime_picker_formfield 2.0.0 [flutter intl]
- equatable 2.0.2 [collection meta]
- firebase_analytics 8.1.1 [firebase_analytics_platform_interface firebase_analytics_web firebase_core flutter meta]
- firebase_auth 1.3.0 [firebase_auth_platform_interface firebase_auth_web firebase_core firebase_core_platform_interface flutter meta]
- firebase_core 1.2.1 [firebase_core_platform_interface firebase_core_web flutter meta]
- firebase_in_app_messaging 0.5.0+5 [firebase_core flutter meta]
- firebase_performance 0.7.0+4 [firebase_core flutter]
- firebase_remote_config 0.10.0+1 [firebase_core firebase_core_platform_interface firebase_remote_config_platform_interface flutter]
- flutter 0.0.0 [characters collection meta typed_data vector_math sky_engine]
- flutter_bloc 7.0.1 [flutter bloc provider]
- flutter_local_notifications 6.0.0 [flutter flutter_local_notifications_platform_interface platform timezone]
- flutter_localizations 0.0.0 [flutter intl characters clock collection meta path typed_data vector_math]
- flutter_native_timezone 2.0.0 [flutter_web_plugins flutter js]
- flutter_secure_storage 4.2.0 [meta flutter]
- flutter_svg 0.22.0 [flutter meta path_drawing vector_math xml]
- flutter_web_browser 0.14.0 [flutter]
- get_it 7.1.3 [async collection]
- google_ml_vision 0.0.5 [flutter]
- holding_gesture 1.1.0 [flutter]
- local_auth 1.1.6 [flutter flutter_plugin_android_lifecycle intl meta platform]
- logging 1.0.1
- medapp_firestore 2.0.0 [flutter dartz get_it pharmacy_core cloud_firestore firebase_core]
- moor 4.3.2 [async convert collection meta pedantic stream_channel sqlite3]
- notificare_push_lib 2.6.0 [flutter json_annotation flutter_plugin_android_lifecycle]
- oauth2 2.0.0 [collection crypto http http_parser]
- package_info 2.0.2 [flutter]
- path_provider 2.0.2 [flutter path_provider_platform_interface path_provider_macos path_provider_linux path_provider_windows]
- percent_indicator 3.0.1 [flutter]
- permission_handler 8.0.1 [flutter meta permission_handler_platform_interface]
- pharmacy_core 1.1.0 [flutter equatable flutter_bloc provider flutter_markdown intl_utils get_it firebase_crashlytics firebase_core]
- provider 5.0.0 [collection flutter nested]
- shared_preferences 2.0.6 [meta flutter shared_preferences_platform_interface shared_preferences_linux shared_preferences_macos shared_preferences_web shared_preferences_windows]
- sliver_tools 0.2.4 [flutter]
- sms_autofill 2.0.0 [pin_input_text_field flutter]
- sqflite_sqlcipher 2.0.0 [flutter synchronized path sqflite_common]
- timeago 3.0.2
- url_launcher 6.0.6 [flutter meta url_launcher_linux url_launcher_macos url_launcher_platform_interface url_launcher_web url_launcher_windows]

dev dependencies:
- bloc_test 8.0.2 [bloc meta mocktail test]
- build_runner 2.0.4 [args async analyzer build build_config build_daemon build_resolvers build_runner_core code_builder collection crypto dart_style frontend_server_client glob graphs http_multi_server io js logging meta mime package_config path pedantic pool pub_semver pubspec_parse shelf shelf_web_socket stack_trace stream_transform timing watcher web_socket_channel yaml]
- chopper_generator 4.0.1 [analyzer build built_collection chopper code_builder dart_style logging meta source_gen]
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters charcode collection matcher meta source_span stream_channel string_scanner term_glyph typed_data]
- mockito 5.0.9 [analyzer build code_builder collection dart_style matcher meta path source_gen test_api]
- moor_generator 4.3.1 [charcode collection recase meta path json_annotation stream_transform args logging cli_util moor sqlite3 sqlparser analyzer analyzer_plugin_fork source_span package_config pub_semver build build_resolvers build_config dart_style source_gen]

dependency overrides:
- encrypted_moor 1.0.0 [moor sqflite_sqlcipher]
- sqlcipher_flutter_libs 0.5.0 [flutter]

transitive dependencies:
- _fe_analyzer_shared 22.0.0 [meta]
- analyzer 1.7.1 [_fe_analyzer_shared cli_util collection convert crypto glob meta package_config path pub_semver source_span watcher yaml pedantic]
- analyzer_plugin_fork 0.5.1 [analyzer dart_style meta pub_semver]
- archive 3.1.2 [crypto path]
- args 2.1.1
- async 2.6.1 [meta collection]
- bloc 7.0.0 [meta]
- boolean_selector 2.1.0 [source_span string_scanner]
- build 2.0.2 [analyzer async convert crypto glob logging meta path]
- build_config 1.0.0 [checked_yaml json_annotation path pubspec_parse yaml]
- build_daemon 3.0.0 [built_collection built_value http_multi_server logging pedantic path pool shelf shelf_web_socket stream_transform watcher web_socket_channel]
- build_resolvers 2.0.3 [analyzer async build crypto graphs logging path package_config pool pub_semver stream_transform yaml]
- build_runner_core 7.0.0 [async build build_config build_resolvers collection convert crypto glob graphs json_annotation logging meta path package_config pedantic pool timing watcher yaml]
- built_collection 5.0.0
- built_value 8.0.6 [built_collection collection fixnum]
- camera_platform_interface 2.0.1 [flutter meta plugin_platform_interface cross_file stream_transform]
- characters 1.1.0
- charcode 1.2.0
- checked_yaml 2.0.1 [json_annotation source_span yaml]
- cli_util 0.3.0 [meta path]
- clock 1.1.0
- cloud_firestore 2.2.1 [cloud_firestore_platform_interface cloud_firestore_web firebase_core firebase_core_platform_interface flutter meta]
- cloud_firestore_platform_interface 5.1.1 [collection firebase_core flutter meta plugin_platform_interface]
- cloud_firestore_web 2.1.1 [cloud_firestore_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins js]
- code_builder 4.0.0 [built_collection built_value collection matcher meta]
- collection 1.15.0
- connectivity_plus_linux 1.0.3 [flutter connectivity_plus_platform_interface dbus meta]
- connectivity_plus_macos 1.0.2 [connectivity_plus_platform_interface flutter]
- connectivity_plus_platform_interface 1.0.2 [flutter meta plugin_platform_interface]
- connectivity_plus_web 1.0.2 [connectivity_plus_platform_interface flutter_web_plugins flutter]
- connectivity_plus_windows 1.0.2 [connectivity_plus_platform_interface flutter]
- convert 3.0.0 [typed_data]
- coverage 1.0.3 [args logging package_config path source_maps stack_trace vm_service]
- cross_file 0.3.1+1 [flutter meta]
- crypto 3.0.1 [collection typed_data]
- dart_style 2.0.1 [analyzer args path pub_semver source_span]
- dbus 0.5.2 [args meta pedantic xml]
- fake_async 1.2.0 [clock collection]
- ffi 1.1.1
- file 6.0.0 [meta path]
- firebase 9.0.1 [http http_parser js]
- firebase_analytics_platform_interface 2.0.1 [flutter meta]
- firebase_analytics_web 0.3.0+1 [firebase firebase_analytics_platform_interface flutter flutter_web_plugins meta]
- firebase_auth_platform_interface 4.2.4 [firebase_core flutter meta plugin_platform_interface]
- firebase_auth_web 1.2.0 [firebase_auth_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins http_parser intl js meta]
- firebase_core_platform_interface 4.0.1 [collection flutter meta plugin_platform_interface]
- firebase_core_web 1.1.0 [firebase_core_platform_interface flutter flutter_web_plugins js meta]
- firebase_crashlytics 2.0.5 [firebase_core firebase_core_platform_interface firebase_crashlytics_platform_interface flutter stack_trace]
- firebase_crashlytics_platform_interface 3.0.5 [collection firebase_core flutter meta plugin_platform_interface]
- firebase_remote_config_platform_interface 0.3.0+1 [firebase_core flutter meta plugin_platform_interface]
- fixnum 1.0.0
- flutter_local_notifications_platform_interface 4.0.0 [flutter plugin_platform_interface]
- flutter_markdown 0.6.2 [flutter markdown meta path]
- flutter_plugin_android_lifecycle 2.0.2 [flutter]
- flutter_web_plugins 0.0.0 [flutter js characters collection meta typed_data vector_math]
- frontend_server_client 2.1.0 [async path]
- glob 2.0.1 [async collection file path pedantic string_scanner]
- graphs 2.0.0
- http 0.13.3 [async http_parser meta path pedantic]
- http_multi_server 3.0.1 [async]
- http_parser 4.0.0 [charcode collection source_span string_scanner typed_data]
- intl 0.17.0 [clock path]
- intl_utils 2.3.0 [analyzer archive args dart_style http intl path petitparser yaml]
- io 1.0.0 [meta path string_scanner]
- js 0.6.3
- json_annotation 4.0.1
- markdown 4.0.0 [args charcode meta]
- matcher 0.12.10 [stack_trace]
- meta 1.3.0
- mime 1.0.0
- mocktail 0.1.1 [collection matcher test]
- nested 1.0.0 [flutter]
- node_preamble 2.0.0
- package_config 2.0.0 [path]
- path 1.8.0
- path_drawing 0.5.1 [vector_math meta path_parsing flutter]
- path_parsing 0.2.1 [vector_math meta]
- path_provider_linux 2.0.0 [path xdg_directories path_provider_platform_interface flutter]
- path_provider_macos 2.0.0 [flutter]
- path_provider_platform_interface 2.0.1 [flutter meta platform plugin_platform_interface]
- path_provider_windows 2.0.1 [path_provider_platform_interface meta path flutter ffi win32]
- pedantic 1.11.0
- permission_handler_platform_interface 3.5.1 [flutter meta plugin_platform_interface]
- petitparser 4.1.0 [meta]
- pin_input_text_field 4.1.0 [flutter]
- platform 3.0.0
- plugin_platform_interface 2.0.0 [meta]
- pool 1.5.0 [async stack_trace]
- process 4.0.0 [file path platform]
- pub_semver 2.0.0 [collection]
- pubspec_parse 1.0.0 [checked_yaml collection json_annotation pub_semver yaml]
- quiver 3.0.1 [matcher]
- recase 3.0.1
- shared_preferences_linux 2.0.0 [flutter file meta path path_provider_linux shared_preferences_platform_interface]
- shared_preferences_macos 2.0.0 [shared_preferences_platform_interface flutter]
- shared_preferences_platform_interface 2.0.0 [flutter]
- shared_preferences_web 2.0.0 [shared_preferences_platform_interface flutter flutter_web_plugins meta]
- shared_preferences_windows 2.0.0 [shared_preferences_platform_interface flutter file meta path path_provider_platform_interface path_provider_windows]
- shelf 1.1.4 [async collection http_parser path stack_trace stream_channel]
- shelf_packages_handler 3.0.0 [path shelf shelf_static]
- shelf_static 1.0.0 [convert http_parser mime path shelf]
- shelf_web_socket 1.0.1 [shelf stream_channel web_socket_channel]
- sky_engine 0.0.99
- source_gen 1.0.1 [analyzer async build dart_style glob meta path pedantic source_span]
- source_map_stack_trace 2.1.0 [path stack_trace source_maps]
- source_maps 0.10.10 [source_span]
- source_span 1.8.1 [collection path term_glyph]
- sqflite_common 2.0.0+2 [synchronized path meta]
- sqlite3 1.1.2 [collection ffi meta]
- sqlparser 0.16.0 [meta collection source_span charcode]
- stack_trace 1.10.0 [path]
- stream_channel 2.1.0 [async]
- stream_transform 2.0.0
- string_scanner 1.1.0 [charcode source_span]
- synchronized 3.0.0
- term_glyph 1.2.0
- test 1.16.8 [analyzer async boolean_selector collection coverage http_multi_server io js node_preamble package_config path pedantic pool shelf shelf_packages_handler shelf_static shelf_web_socket source_span stack_trace stream_channel typed_data web_socket_channel webkit_inspection_protocol yaml test_api test_core]
- test_api 0.3.0 [async boolean_selector collection meta path source_span stack_trace stream_channel string_scanner term_glyph matcher]
- test_core 0.3.19 [analyzer async args boolean_selector collection coverage glob io meta package_config path pedantic pool source_map_stack_trace source_maps source_span stack_trace stream_channel vm_service yaml matcher test_api]
- timezone 0.7.0 [path]
- timing 1.0.0 [json_annotation]
- typed_data 1.3.0 [collection]
- url_launcher_linux 2.0.0 [flutter]
- url_launcher_macos 2.0.0 [flutter]
- url_launcher_platform_interface 2.0.3 [flutter plugin_platform_interface]
- url_launcher_web 2.0.1 [flutter flutter_web_plugins meta url_launcher_platform_interface]
- url_launcher_windows 2.0.0 [flutter]
- vector_math 2.1.0
- vm_service 6.2.0
- watcher 1.0.0 [async path pedantic]
- web_socket_channel 2.1.0 [async crypto stream_channel]
- webkit_inspection_protocol 1.0.0 [logging]
- win32 2.1.3 [ffi]
- xdg_directories 0.2.0 [meta path process]
- xml 5.1.1 [collection meta petitparser]
- yaml 3.1.0 [collection source_span string_scanner]

```

</details>

---

**Comment by markusaksli-nc** [comment]
Hi @devaronius 
Could you please provide a [minimal complete reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example)?
Thank you

**Comment by HeavenOSK** [comment]
Hi, @markusaksli-nc 

I have reproduced this issue in my project. 
I've created the minimal code for you to refer to the following code.
 
https://github.com/HeavenOSK/save_document_reference_error/blob/main/lib/main.dart

**Comment by HeavenOSK** [comment]
workaround: 

If we don't use `withConverter`, we can save `DocumentReference` field.
```dart
final docRef = firestore.collection('someCollection').doc();  
final data = SomeDocument(docRef: docRef);  
await docRef.set(    
  SomeDocument.toJson(data, null),  
);
```
https://github.com/HeavenOSK/save_document_reference_error/blob/workaround/lib/main.dart

**Comment by devaronius** [comment]
@HeavenOSK  thanks for the small minimal code, and found the Workaround as well. 
But that should not be necessary that it tries to run the withConverter when you only ask for the reference

**Comment by markusaksli-nc** [comment]
Seems reproducible with the provided code.
```yaml
environment:
  sdk: ">=2.12.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  firebase_core: ^1.3.0
  cloud_firestore: ^2.2.2
```
<details>
 <summary>flutter doctor -v</summary>

```
[√] Flutter (Channel master, 2.3.0-17.0.pre.504, on Microsoft Windows [Version 10.0.18363.1316], locale et-EE)
    • Flutter version 2.3.0-17.0.pre.504 at C:\Development\flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 102acbe39d (2 hours ago), 2021-06-24 22:51:02 -0700
    • Engine revision 4bf9d638c2
    • Dart version 2.14.0 (build 2.14.0-245.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at C:\Users\Isa\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: C:\Users\Isa\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\202.7351085\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.6.4)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.6.30309.148
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.2.0)
    • Android Studio at C:\Users\Isa\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\202.7351085
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[√] VS Code (version 1.47.2)
    • VS Code at C:\Users\Isa\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.12.2

[√] Connected device (2 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.18363.1316]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 85.0.4183.121

• No issues found!
```
</details>

**Comment by JPGarCar** [comment]
Any updates on a fix for this issue?

**Comment by JPGarCar** [comment]
I am getting this same issue when doing where queries with a DocumentReference<T> and when trying to set the DocumentReference<T> on a document field of type reference.

My app is not usable at this point. Is this going to be looked at any time soon?

**Comment by devaronius** [comment]
@JPGarCar can you share your fix here? or is this included in the library?

**Comment by JPGarCar** [comment]
@devaronius It's not included in the library. I created a few FirestoreCollection and FirestoreDocument objects to refactor some code. These objects have a clean collection reference from which I can get a clean document reference (DocumentReference<Map<String, dynamic>>). You can take a look at my code [here](https://github.com/JPGarCar/stw_timing_web_app/tree/main/lib/models/firestore/objects). Take a look at firestore_collection_object.dart, last get function.

**Comment by tomoyuki28jp** [comment]
I am getting this same issue :(

**Comment by nico151999** [comment]
I've also encountered the reported bug. As the suggested workarounds were no option for me, I decided to take a look at it and created a pull request. I am not as much into Flutter as I am into other technologies so I do not want to promise you to have followed all best practices but the changes were rather minor which keeps me hoping for a quick merge. Let's wait for the required two approving reviews. Fingers crossed :crossed_fingers: :smiley:
